build: fix build fail on clang 15

Totktonada opened the pull request:
Fixed pthread-related CMake checks. The checks code is built with `-pedantic-errors` and it leads to errors of the following kind on clang 15:

```c
<...>/CMakeFiles/CMakeScratch/TryCompile-78KaOK/src.c:4:17: error: a
    function declaration without a prototype is deprecated in all
    versions of C [-Werror,-Wstrict-prototypes]
        int main() { pthread_setname_np(pthread_self(), ""); }
                ^
                 void
```

Fixed a warning in the SQL code (it's an error in Debug build):

```c
<...>/src/box/sql/vdbeaux.c:170:13: error: variable 'n' set but not used
    [-Werror,-Wunused-but-set-variable]
        static int n = 0;
```

Fixed several warnings from lemon.c of the following kind:

```c
<...>/extra/lemon.c:173:6: warning: a function declaration without a
    prototype is deprecated in all versions of C and is treated as a
    zero-parameter prototype in C2x, conflicting with a subsequent
    definition [-Wdeprecated-non-prototype]
void FindRulePrecedences();
     ^
<...>/extra/lemon.c:766:6: note: conflicting prototype is here
void FindRulePrecedences(struct lemon *xp)
```

See also https://github.com/tarantool/small/issues/57

Fixes #8110

CuriousGeorgiy reviewed (changes requested):
Grepping the sources for `int main()` I found at least these files not fixed: 
* cmake/simd.cmake;
* cmake/FindCurl.cmake;
* test/unit/say.c;
* test/unit/queue.c;
* etc.

If you grep for `()\n{` in `*.c` files, you will also find a number of places that, AFAIC, should be diagnosed (i.e., they need to be replaced with `(void)\n{`.

Just to clarify, did you build all the targets?

Totktonada commented:
> Grepping the sources for `int main()` I found at least these files not fixed: 
> <...>
>
> If you grep for `()\n{` in `*.c` files, you will also find a number of places that, AFAIC, should be diagnosed (i.e., they need to be replaced with `(void)\n{`).

I found that the reason of the error is `-pedantic-errors`, which is set in `cmake/thread.cmake`. Just `-Wall -Wextra -Werror` doesn't issue any warning/error. So I removed all the changes aside of the `cmake/thread.cmake` one.

However I also tried the debug build, found several new warnings and added changes to SQL and SQL parser code. Also found an unused variable in a libsmall test and opened https://github.com/tarantool/small/pull/56.

> I don't have clang 15 on hand, so, just to clarify, did you build all the targets?

I call `cmake <...> && make -j`. Should I check anything else?

CuriousGeorgiy commented:
> I call cmake <...> && make -j. Should I check anything else?

Yeah, that should be enough, both Release and Debug builds.

Totktonada commented:
> Nit: I would rather open a ticket in tarantool/small than refer to a PR — or I would first merge the PR to tarantool/small and refer to the correpsonding commit.

Please, let me know if I should performs some actions.

> I think we should either drop the `-pedantic-errors` flag from cmake/thread.cmake or enable it for the entire build. Not going to block the patch, but I would like to hear @alyapunov's opinion.

It was added in cf8d8553084894769c2629719b8e6c92572c9092. As I see, we need it to escalate the implicit function declaration warning to the error. However `-Werror` would work here too.

Healing the checks looks better for me, but I don't mind any option.

> Yeah, that should be enough, both Release and Debug builds.

I verified RelWithDebInfo and Debug. We don't use the Release build, because tarantool leans on presence of the debug information for generating backtraces in `fiber.info()` and at crash.

Totktonada commented:
> > Nit: I would rather open a ticket in tarantool/small than refer to a PR — or I would first merge the PR to tarantool/small and refer to the correpsonding commit.
>
> Please, let me know if I should performs some actions.

Filed https://github.com/tarantool/small/issues/57, linked from here.

Totktonada commented:
I'll push this patch after https://github.com/tarantool/small/pull/56. Waiting for now.

Totktonada commented:
libsmall was updated in:

* `master` in PR #8142.
* `2.10` in PR #8143.
* `1.10` in PR #8144.

I'm going to proceed with this patch now.

Totktonada commented:
Pushed to:

* `master` in 1c6b6f85f4d36787f9e0a861a5437412e65bae11 (2.11.0-entrypoint-860-g1c6b6f85f)
* `2.10` in 26721cbb0f21f35b0b0c00abdc30575e89ef7cfd (2.10.4-50-g26721cbb0)
* `1.10` in 20a0789c6f71d0803e2fd143def270d43c14471f (1.10.14-52-g20a0789c6)
